File: src/converter/pos_util.cc
// Copyright 2010, Google Inc.
// All rights reserved.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are
// met:
//
//     * Redistributions of source code must retain the above copyright
// notice, this list of conditions and the following disclaimer.
//     * Redistributions in binary form must reproduce the above
// copyright notice, this list of conditions and the following disclaimer
// in the documentation and/or other materials provided with the
// distribution.
//     * Neither the name of Google Inc. nor the names of its
// contributors may be used to endorse or promote products derived from
// this software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

#include <algorithm>
#include <string>
#include <vector>
#include "base/base.h"
#include "base/util.h"
#include "base/file_stream.h"
#include "converter/pos_util.h"

namespace mozc {

void POSUtil::Open(const string &id_file) {
  ids_.clear();
  InputFileStream ifs(id_file.c_str());
  CHECK(ifs);
  string line;
  vector<string> fields;
  while (getline(ifs, line)) {
    if (line.empty() || line[0] == '#') {
      continue;
    }
    fields.clear();
    Util::SplitStringUsing(line, "\t ", &fields);
    CHECK_GE(fields.size(), 2);
    const int id = atoi32(fields[0].c_str());
    ids_.push_back(make_pair(fields[1], static_cast<uint16>(id)));
  }

  //  const char kNumberPOS[] = "名詞,数";
  const char kNumberPOS[] = "\xE5\x90\x8D\xE8\xA9\x9E,\xE6\x95\xB0";
  CHECK(ids(kNumberPOS, &number_ids_));
  sort(number_ids_.begin(), number_ids_.end());

  // "助詞"
  const char kParticlePOS[] = "\xe5\x8a\xa9\xe8\xa9\x9e";
  // "助動詞"
  const char kAuxVerbPOS[] = "\xe5\x8a\xa9\xe5\x8b\x95\xe8\xa9\x9e";
  // "記号"
  const char kSymbolPOS[] = "\xe8\xa8\x98\xe5\x8f\xb7";
  // "動詞,非自立"
  const char kVerbDependentPOS[] =
      "\xE5\x8B\x95\xE8\xA9\x9E,\xE9\x9D\x9E\xE8\x87\xAA\xE7\xAB\x8B";
  // "名詞,非自立"
  const char kNounDependentPOS[] =
      "\xE5\x90\x8D\xE8\xA9\x9E,\xE9\x9D\x9E\xE8\x87\xAA\xE7\xAB\x8B";
  // "形容詞,非自立"
  const char kAdjectiveDependentPOS[] =
      "\xE5\xBD\xA2\xE5\xAE\xB9\xE8\xA9\x9E,\xE9\x9D\x9E\xE8\x87\xAA\xE7\xAB\x8B";
  // "動詞,接尾"
  const char kVerbSuffixPOS[] =
      "\xE5\x8B\x95\xE8\xA9\x9E,\xE6\x8E\xA5\xE5\xB0\xBE";
  // "名詞,接尾"
  const char kNounSuffixPOS[] =
      "\xE5\x90\x8D\xE8\xA9\x9E,\xE6\x8E\xA5\xE5\xB0\xBE";
  // "形容詞,接尾"
  const char kAdjectiveSuffixPOS[] =
      "\xE5\xBD\xA2\xE5\xAE\xB9\xE8\xA9\x9E,\xE6\x8E\xA5\xE5\xB0\xBE";

  CHECK(ids(kParticlePOS, &functional_word_ids_));
  CHECK(ids(kAuxVerbPOS, &functional_word_ids_));
  CHECK(ids(kSymbolPOS, &functional_word_ids_));
  CHECK(ids(kVerbDependentPOS, &functional_word_ids_));
  CHECK(ids(kNounDependentPOS, &functional_word_ids_));
  CHECK(ids(kAdjectiveDependentPOS, &functional_word_ids_));
  CHECK(ids(kVerbSuffixPOS, &functional_word_ids_));
  CHECK(ids(kNounSuffixPOS, &functional_word_ids_));
  CHECK(ids(kAdjectiveSuffixPOS, &functional_word_ids_));

  sort(functional_word_ids_.begin(), functional_word_ids_.end());
}

uint16 POSUtil::id(const string &feature) const {
  CHECK(!feature.empty());
  for (size_t i = 0; i < ids_.size(); ++i) {
    if (ids_[i].first.find(feature) == 0) {
      return ids_[i].second;
    }
  }
  LOG(ERROR) << "Cannot find the POS for: " << feature;
  return 0;
}

bool POSUtil::ids(const string &feature,
                  vector<uint16> *ids) const {
  CHECK(ids);
  CHECK(!feature.empty());
  bool found = false;
  for (size_t i = 0; i < ids_.size(); ++i) {
    if (ids_[i].first.find(feature) == 0) {
      ids->push_back(ids_[i].second);
      found = true;
    }
  }
  LOG_IF(ERROR, !found) << "Cannot find the POS for: "
                        << feature;
  return found;
}
}  // namespace mozc

File: src/dictionary/pos_util.py
# -*- coding: utf-8 -*-
# Copyright 2010-2021, Google Inc.
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are
# met:
#
#     * Redistributions of source code must retain the above copyright
# notice, this list of conditions and the following disclaimer.
#     * Redistributions in binary form must reproduce the above
# copyright notice, this list of conditions and the following disclaimer
# in the documentation and/or other materials provided with the
# distribution.
#     * Neither the name of Google Inc. nor the names of its
# contributors may be used to endorse or promote products derived from
# this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

"""Utilities to handle pos related stuff for source code generation."""

import codecs
import collections
import logging
import re

from build_tools import code_generator_util


class PosDataBase(object):
  """Utility to look up data in id.def and special_pos.def."""

  def __init__(self):
    self.id_list = []

  def Parse(self, id_file, special_pos_file):
    id_list = []
    with codecs.open(id_file, 'r', encoding='utf-8') as stream:
      stream = code_generator_util.SkipLineComment(stream)
      stream = code_generator_util.ParseColumnStream(stream, num_column=2)
      for pos_id, feature in stream:
        id_list.append((feature, int(pos_id)))

    max_id = max(pos_id for _, pos_id in id_list)
    with codecs.open(special_pos_file, 'r', encoding='utf-8') as stream:
      stream = code_generator_util.SkipLineComment(stream)
      for pos_id, line in enumerate(stream, start=max_id + 1):
        id_list.append((line, pos_id))
    self.id_list = id_list

  def GetPosId(self, feature):
    """Returns id for the feature if found. Otherwise None."""
    assert feature
    for line, pos_id in self.id_list:
      # Return by prefix match.
      if line.startswith(feature):
        return pos_id

    logging.warning('Cannot find the POS for: %s', feature)

  @staticmethod
  def _GroupConsecutiveId(iterable):
    result = []
    for value in iterable:
      if result and result[-1] + 1 != value:
        yield result
        result = []
      result.append(value)
    if result:
      yield result

  def GetRange(self, pattern):
    id_list = [pos_id for line, pos_id in self.id_list if pattern.match(line)]
    id_list.sort()
    return [
        (id_range[0], id_range[-1])
        for id_range in PosDataBase._GroupConsecutiveId(id_list)
    ]


class PosMatcher(object):

  def __init__(self, pos_database):
    self.pos_database = pos_database
    self._match_rule_map = {}

  def Parse(self, pos_matcher_rule_file):
    with codecs.open(pos_matcher_rule_file, 'r', encoding='utf-8') as stream:
      stream = code_generator_util.SkipLineComment(stream)
      stream = code_generator_util.ParseColumnStream(stream, num_column=2)
      self._match_rule_map = dict(
          (name, (pattern, re.compile(pattern.replace('*', '[^,]+')), sortkey))
          for sortkey, (name, pattern) in enumerate(stream)
      )

  def GetRuleNameList(self):
    """Returns a list of rule names in the original file's order."""
    sorted_rule_list = sorted(
        list(self._match_rule_map.items()), key=lambda item: item[1][2]
    )
    return [rule_name for rule_name, _ in sorted_rule_list]

  def GetRange(self, name):
    return self.pos_database.GetRange(self._match_rule_map[name][1])

  def GetId(self, name):
    return self.pos_database.GetRange(self._match_rule_map[name][1])[0][0]

  def GetOriginalPattern(self, name):
    return self._match_rule_map[name][0]


class InflectionMap(object):
  """Utility to handle inflection map.

  Inflection map is a map from key to (form, value_suffix, key_suffix).
  """

  def __init__(self):
    self._map = {}

  def Parse(self, filepath):
    result = collections.defaultdict(list)
    with codecs.open(filepath, 'r', encoding='utf-8') as stream:
      stream = code_generator_util.SkipLineComment(stream)
      stream = code_generator_util.ParseColumnStream(stream, num_column=4)
      for key, form, value_suffix, key_suffix in stream:
        result[key].append((
            form,
            value_suffix if value_suffix != '*' else '',
            key_suffix if key_suffix != '*' else '',
        ))
    self._map = result

  def Get(self, key):
    return self._map[key]


class UserPos(object):
  """Utility to handle user pos.

  The data is assoc list from user_pos (string) to conjugation_list.
  conjugation_list is a list of (value_suffix, key_suffix, pos_id).
  """

  def __init__(self, pos_database, inflection_map):
    self._pos_database = pos_database
    self._inflection_map = inflection_map
    self.data = []

  def Parse(self, filepath):
    result = []
    with codecs.open(filepath, 'r', encoding='utf-8') as stream:
      stream = code_generator_util.SkipLineComment(stream)
      stream = code_generator_util.ParseColumnStream(stream, num_column=4)
      for user_pos, _, ctype, feature in stream:
        conjugation_list = []
        if ctype == '*':
          conjugation_list.append(
              (None, None, self._pos_database.GetPosId(feature))
          )
        else:
          for form, value_suffix, key_suffix in self._inflection_map.Get(ctype):
            # repalce <cfrom> with actual cform
            pos_id = self._pos_database.GetPosId(
                feature.replace('<cform>', form)
            )

            # Known error items.
            # 動詞,自立,*,*,五段動詞,体言接続特殊２,*
            # 形容詞,自立,*,*,形容詞・アウオ段,文語基本形,*
            if pos_id is not None:
              conjugation_list.append((value_suffix, key_suffix, pos_id))

        result.append((user_pos, conjugation_list))
    self.data = result

  def GetPosId(self, pos):
    """Returns id of the given pos."""
    for user_pos, conjugation_list in self.data:
      if user_pos == pos:
        return conjugation_list[0][2]
